File: myviews/src/main/java/pl/mareklangiewicz/myviews/MyViewDecorator.java
package pl.mareklangiewicz.myviews;

import android.support.annotation.LayoutRes;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import java.util.Map;

/**
 * Created by Marek Langiewicz on 31.05.15.
 * TODO later: javadocs and tests
 * We do not decorate inside already decorated view.
 */
public final class MyViewDecorator {

    static public void decorateOne(@NonNull View view, @LayoutRes int decoration, @Nullable LayoutInflater inflater) {

        if(inflater == null)
            inflater = LayoutInflater.from(view.getContext());

        ViewGroup.LayoutParams params = view.getLayoutParams();
        ViewGroup parent = (ViewGroup) view.getParent();
        int idx = parent.indexOfChild(view);
        parent.removeViewAt(idx);
        View decorated = inflater.inflate(decoration, null);
        parent.addView(decorated, idx, params);
        ViewGroup content = (ViewGroup) decorated.findViewById(android.R.id.content);
        content.addView(view,
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.MATCH_PARENT);

    }

    static public void decorateTree(@NonNull View root, @NonNull String tag, @LayoutRes int decoration,
                                    @Nullable LayoutInflater inflater) {
        if(inflater == null)
            inflater = LayoutInflater.from(root.getContext());

        if(tag.equals(root.getTag())) {
            decorateOne(root, decoration, inflater);
        }
        else if(root instanceof ViewGroup) {
            ViewGroup group = (ViewGroup) root;
            int N = group.getChildCount();
            for(int i = 0; i < N; ++i) {
                decorateTree(group.getChildAt(i), tag, decoration, inflater);
            }
        }
    }

    static public void decorateTree(@NonNull View root, Map<String, Integer> decorations,
                                    @Nullable LayoutInflater inflater) {
        if(inflater == null)
            inflater = LayoutInflater.from(root.getContext());
        if(decorations.containsKey(root.getTag().toString())) {
            decorateOne(root, decorations.get(root.getTag().toString()), inflater);
        }
        else if(root instanceof ViewGroup) {
            ViewGroup group = (ViewGroup) root;
            int N = group.getChildCount();
            for(int i = 0; i < N; ++i) {
                decorateTree(group.getChildAt(i), decorations, inflater);
            }
        }
    }

}

File: myviews/src/main/java/pl/mareklangiewicz/myviews/MyViewDecorator.kt
package pl.mareklangiewicz.myviews

import android.support.annotation.LayoutRes
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.ViewGroup.LayoutParams.MATCH_PARENT

/**
 * Created by Marek Langiewicz on 31.05.15.
 * This class contains some static methods that allow to inject some decorations
 * to selected views in view hierarchy. It is probably bad idea to mess with
 * default view hierarchy generation, so... don't use this class. :-)
 * We do not decorate inside already decorated view.
 */
object MyViewDecorator {

        fun decorateOne(view: View, @LayoutRes decoration: Int, inflater: LayoutInflater = LayoutInflater.from(view.context)) {
            val params = view.layoutParams
            val parent = view.parent as ViewGroup
            val idx = parent.indexOfChild(view)
            parent.removeViewAt(idx)
            val decorated = inflater.inflate(decoration, null)
            parent.addView(decorated, idx, params)
            val content = decorated.findViewById(android.R.id.content) as ViewGroup
            content.addView(view, MATCH_PARENT, MATCH_PARENT)
        }

        fun decorateTree(root: View, tag: String, @LayoutRes decoration: Int, inflater: LayoutInflater = LayoutInflater.from(root.context)) {
            if (tag == root.tag) decorateOne(root, decoration, inflater)
            else if (root is ViewGroup)
                for (i in 0..root.childCount - 1)
                    decorateTree(root.getChildAt(i), tag, decoration, inflater)
        }

        fun decorateTree(root: View, decorations: Map<String, Int>, inflater: LayoutInflater = LayoutInflater.from(root.context)) {
            val d = decorations[root.tag.toString()]
            if (d !== null)
                decorateOne(root, d, inflater)
            else if (root is ViewGroup)
                for (i in 0..root.childCount - 1)
                    decorateTree(root.getChildAt(i), decorations, inflater)
        }
}
